pyface/ui/qt4/widget.py

```diff
@@ -105,6 +105,10 @@ def __event_filter_default(self):
 class WidgetEventFilter(QtCore.QObject):
     """ An internal class that watches for certain events on behalf of the
     Widget instance.
+
+    This filter watches for show and hide events to make sure that visible
+    state of the widget is the opposite of Qt's isHidden() state.  This is
+    needed in case other code hides the toolkit widget
```

Comment:
I believe the `visible` trait is meant for toggling, not for query.
e.g. one can toggle visibility like this:
```
def toggle_visibility(widget):
    widget.visible = not widget.visible
```

But if one wants to know whether a widget is actually visible by the user, this is not the trait to use.
https://github.com/enthought/pyface/blob/e9b7d968f5fbfb239d70c4c1e095e0672aa0b458/pyface/i_widget.py#L29-L30

Just reading the documentation, I'd have thought the trait can also be used for querying the visibility state.

While using the attribute name `visible` to mean `not hidden` might be convenient for the toggling use case, it will preclude the addition of a separate attribute that actually reflects visibility should a new use case for query arise.  Shall we just add a `hidden` trait to decouple them?  Then the `visible` trait can be implemented like a property, where its getter reflects the value from Qt `isVisible`.

Reply:
For better or worse, `visible` is tracking the opposite of what Qt calls `hidden`. As mentioned in the comments for the issue, this is matching the way that `visible` behaves on the Wx backend and the way that TraitsUI's `visible_when` behaves (and I think, although I have not checked, how the `visible` attribute of an `Action` works).

So this is the right change for now.

There may well be room for a read-only trait that can be queried for the "am I shown on the screen right now" state, but for backwards compatibility it should probably not be called `visible`.

Reply:
I did read the wx code but it is not obvious to me that the behaviour in wx is also about toggling.  Maybe it is, but none of the tests here checks the wx states to ensure this is what the trait does, so I am not sure.

If it is intended to behave like `visible_when`, then the `visible` is a toggle for the developers to change the widget visibility. It should not be updated via Qt event right?

If I meant for the widget to be hidden, then `widget.visible` is false. In that case, `if event_type == QtCore.QEvent.Show` should not occur, so the first branch is not used.  In fact, if I remove this branch, none of the tests fail.

I think the code would be clearer if we have this instead:
```
if event_type == QtCore.QEvent.Hide:
    widge.visible = not widget.control.isHidden()
```
Then it is more straightforward `visible` is simply `not hidden`. If I make this change, all the tests (include the ones I proposed) still pass.

We should also update the documentation for `visible` to indicate that this trait cannot be used for query, otherwise it is tempting to misuse it:
https://github.com/enthought/pyface/blob/e9b7d968f5fbfb239d70c4c1e095e0672aa0b458/pyface/i_widget.py#L29-L30

Reply:
> There may well be room for a read-only trait that can be queried for the "am I shown on the screen right now" state, but for backwards compatibility it should probably not be called visible.

Sigh, I don't know what that new trait should be called if `visible` is already taken here. My only hope is that it won't be called `hidden`, because it would just be doing exactly the opposite of Qt, deliberately messing with developers.

Reply:
> if event_type == QtCore.QEvent.Hide:
>     widge.visible = not widget.control.isHidden()

This was my first instinct too, but the problem is that _setting_ (ie. calls to `setVisible()`) has side-effects in Qt, so we should set _only_ when we need to.  I think I tried this and it caused issues with the display of the widget (but I could be mis-remembering the path I went through to come up with this fix - I certainly tried something like this).  Qt

As for another trait, yes I agree `hidden` is a bad name, maybe `displayed` or something like that - and I certainly don't want people getting intuition from the way Qt handles visibility/hidden state.  But that's a discussion for a separate issue.  Note that I haven't ever needed that information at the sort of level that Pyface etc. are working (it can be important for eg. halting video playback to save CPU, but if you were implementing that, you're probably working directly with Qt).

Reply:
> This was my first instinct too, but the problem is that setting (ie. calls to setVisible()) has side-effects in Qt, so we should set only when we need to. I think I tried this and it caused issues with the display of the widget (but I could be mis-remembering the path I went through to come up with this fix - I certainly tried something like this). Qt

That would be weird, setting the `visible` will only cause `setVisible` to be invoked if the trait has changed, right? If we can't simply set the `visible` trait to sync up with Qt, it smells like a bigger problem and is worth looking into.

Reply:
If we cannot reproduce the "it caused issues with the display of the widget", then I'd still prefer doing the obviously right thing: set the `visible` to what Qt thinks it is. If the issue you said exists, we can fix it when it surfaces, knowing what exactly the problem is.  Otherwise it feels like we are working around something we don't exactly know what, and that workaround might introduce new problems, and more workarounds to the point where nothing is fixable.

Reply:
FWIW I think I am going crazy - I simplified the whole thing to one branch:
```
        if event_type in {QtCore.QEvent.Show, QtCore.QEvent.Hide}:
            widget.visible = not widget.control.isHidden()
```
and it seems to work correctly.  I could have sworn that while I was hammering out the fix I tried either the above, or what you were suggesting and there were problems, but I can't see any right now, and tests pass.  There must have been some other orthogonal issue that was messing with things.